NeuroID/src/advancedDeviceLib/java/com/neuroid/tracker/service/AdvancedDeviceIDManagerService.kt

```diff
@@ -0,0 +1,197 @@
+    private val advNetworkService: ADVNetworkService,
+    private val fpjsClient: FingerprintJS? // only for testing purposes, need to create in real time to pass NID Key
+): AdvancedDeviceIDManagerService {
+    companion object {
+        internal val NID_RID = "NID_RID_KEY"
+        internal val defaultCacheValue = "{\"key\":\"NO_KEY\", \"exp\":0}"
+    }
+
+    override fun getCachedID():Boolean{
+        val existingString = sharedPrefs.getString(NID_RID, defaultCacheValue)
+
+        val gson = Gson()
+        val type = object : TypeToken<Map<String, Any>>() {}.type
+        val storedValue = gson.fromJson<Map<String, Any>>(existingString, type)
+
+        // No Value Exists - return false
+        if (storedValue == null || storedValue["key"] == "NO_KEY") {
+            return false
+        }
+
+        // Expired ID - return false
+        val currentTimestamp = System.currentTimeMillis()
+        if (currentTimestamp > storedValue["exp"] as Double) {
+            return false
+        }
+
+        // Capture valid cached ID
+        logger.d(
+            msg = "Retrieving Request ID for Advanced Device Signals from cache: ${storedValue["key"]}"
+        )
+        dataStoreManager.saveEvent(
+            NIDEventModel(
+                type = ADVANCED_DEVICE_REQUEST,
+                rid = storedValue["key"] as String,
+                ts = System.currentTimeMillis(),
+                c = true
+            )
+        )
+        
+        return true
+    }
+
+    override fun getRemoteID(
+        clientKey:String,
+        remoteUrl:String
+    ): Job? {
+        // Retrieving the API key from NeuroID
+        val nidKeyResponse = advNetworkService.getNIDAdvancedDeviceAccessKey(clientKey)
+
+        // if no key exit early
+        if(!nidKeyResponse.success) {
+            logger.e(msg = "Failed to get API key from NeuroID: ${nidKeyResponse.message}")
+
+            // capture failure in event
+            dataStoreManager
+                .saveEvent(
+                    NIDEventModel(
+                        type = LOG,
+                        ts = System.currentTimeMillis(),
+                        level="error",
+                        m=nidKeyResponse.message
+                    )
+                )
+            return null
+        }
+
+        // Call FPJS with our key
+        val fpjsClient = if( fpjsClient != null) {
+            fpjsClient
+        } else {
+            FingerprintJSFactory(applicationContext = context)
+                .createInstance(
+                    Configuration(
+                        apiKey = nidKeyResponse.key,
+                        endpointUrl = remoteUrl
+                    )
+                )
+        }
+
+        val remoteIDJob = CoroutineScope(Dispatchers.IO).launch {
+            var maxRetryCount = NIDAdvancedDeviceNetworkService.RETRY_COUNT
+            var retryCount = 0
+
+            while (retryCount < maxRetryCount) {
```

Comment:
Restructure loop to use intrinsic index:

```suggestion
for (retryCount in 1..maxRetryCount) {
```